Added SshCommand.InputStream to allow writing to stdin of SshCommand

realvizu opened the pull request:
Changes:
- Added SshCommand.InputStream (type: ChannelInputStream), which sends the data written to the stream through a command's channel.
- Added an integration test that echos back the sent data using the "cat" command.
- All unit tests pass.

I you have recommendations for further unit tests, please advise.

---

I started the work by trying to run all unit tests on my dev machine to see if they all pass. Unfortunately a lot of them failed, so I had to make the following modifications (see the changes in my first commit in this PR):
- Excluded MD5 tests for the net462 target, because I got System.InvalidOperationException: 'This implementation is not part of the Windows Platform FIPS validated cryptographic algorithms.'
- Modified SshdConfig.cs: do not throw for "Include", just do nothing.
- Modified failing dos2unix parameters in Dockerfile.TestServer.
- Added a line to .gitattributes to force LF line ending for key files used by integration tests, otherwise using them causes error.
- Modified SftpClientTest.Test_Sftp_Multiple_Async_Upload_And_Download_10Files_5MB_Each to upload only 2 files instead of 10 because for more than 2 files I got timeout errors.

If you know a better fix for any of the above modifications, please advise.

Commit:
Making all unit tests pass locally.
Excluded MD5 tests on net462 because I get System.InvalidOperationException: 'This implementation is not part of the Windows Platform FIPS validated cryptographic algorithms.'
SshdConfig: do not throw for "Include", just do nothing.
Modified failing dos2unix parameters in Dockerfile.TestServer.
Forceing LF line ending for key files used by integration tests, otherwise using them causes error.
SftpClientTest.Test_Sftp_Multiple_Async_Upload_And_Download_10Files_5MB_Each times out for maxFiles=10, decreasing to 2 to make the test pass.

Commit:
Added SshCommand.InputStream.

Commit:
Added an integration test for SshCommand.InputStream.

WojciechNagorski commented on test/Renci.SshNet.IntegrationTests/OldIntegrationTests/SftpClientTest.Upload.cs:
```diff
@@ -78,7 +78,9 @@ public void Test_Sftp_Upload_Forbidden()
-            var maxFiles = 10;
+            // Works for up to 2 files, but fails for more files (on my machine).
+            // I get either "Renci.SshNet.Common.SshException: Channel was closed" or timeout exception.
+            var maxFiles = 2;
```
I'll have to check this on my machine because it seems to be working now. But I know that this test is unstable. so don't worry about it.

Rob-Hague replied:
It also fails on my machine, but I've not gotten to the bottom of why (I think sshd is crashing but maybe SSH.NET is doing something wrong). I think we should revert this change for this PR

WojciechNagorski replied:
Yes, I agree because this is not a part of this PR. @realvizu Can you revert this?

realvizu replied:
Sure, reverted it.

WojciechNagorski commented on test/Renci.SshNet.Tests/Classes/Common/HostKeyEventArgsTest.cs:
```diff
@@ -45,6 +45,10 @@ public void HostKeyEventArgsConstructorTest()
 
+// Excluding on net462 platform, because using MD5 hash throws: System.InvalidOperationException:
+// 'This implementation is not part of the Windows Platform FIPS validated cryptographic algorithms.'
+#if !NET462_TARGET_FRAMEWORK
```
I'll have to check this on my machine because it seems to be working now.

Rob-Hague replied:
Seems like @realvizu has FIPS enabled. We should probably condition these tests for when [CryptoConfig.AllowOnlyFipsAlgorithms](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.cryptoconfig.allowonlyfipsalgorithms) is false. In the meantime, can we revert these changes for this PR?

WojciechNagorski replied:
Yes, I agree because this is not a part of this PR. @realvizu Can you revert this?

realvizu replied:
Sure, reverted it.

WojciechNagorski commented on test/Renci.SshNet.Tests/Classes/PrivateKeyFileTest.cs:
```diff
@@ -1,4 +1,8 @@
-﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+﻿// Excluding these tests on net462 platform, beacuse using MD5 hash throws: System.InvalidOperationException:
+// 'This implementation is not part of the Windows Platform FIPS validated cryptographic algorithms.'
+#if !NET462_TARGET_FRAMEWORK
```
I'll have to check this on my machine because it seems to be working now.

WojciechNagorski replied:
@realvizu Can you revert this? It is not a part of this PR.

realvizu replied:
Sure, reverted it.

WojciechNagorski commented on test/Renci.SshNet.IntegrationTests/Dockerfile.TestServer:
```diff
@@ -10,7 +10,8 @@ RUN apk update && apk upgrade --no-cache && \
     apk add --no-cache openssh-server-pam && \
-    dos2unix /etc/ssh/* && \
+    # must not use * for dos2unix parameter otherwise it tries to process folders too and fails
+    dos2unix /etc/ssh/ssh*key && \
```
Maybe there is a new alpine image and it has different behavior, but I don't have a problem with this change as long as it works.

WojciechNagorski commented:
It looks good to me! But I need more time to test it!

Do you have a specific function in mind that you want to implement using this functionality? 
Did you see that it can be useful in https://github.com/dotnet/Docker.DotNet/issues/540

WojciechNagorski committed:
Merge branch 'develop' into SshCommandInputStream

WojciechNagorski commented:
![image](https://github.com/sshnet/SSH.NET/assets/17333903/7721dd40-b98e-4657-b8a9-a657a236a455)
Those tests work on my machine.

Rob-Hague commented on src/Renci.SshNet/Common/ChannelInputStream.cs:
```diff
@@ -0,0 +1,221 @@
+            _totalPosition += count;
+
+            // Must send EOF, otherwise SshCommand.EndExecute never gets called.
+            _channel.SendEof();
```
This does not look right. What if `Write` is called multiple times? Can another test be added which writes one byte at a time to validate that? It is a valid scenario when calling `stream.CopyTo` which calls `Write` under the hood

realvizu replied:
You're absoloutely right, I moved the EOF sending to Dispose, and added a unit test that writes the input one byte at a time.

Rob-Hague commented on test/Renci.SshNet.TestTools.OpenSSH/SshdConfig.cs:
```diff
@@ -385,6 +385,8 @@ private static void ProcessGlobalOption(SshdConfig sshdConfig, string line)
                 case "PasswordAuthentication":
                 case "GatewayPorts":
+                // Had to add this otherwise docker container setup in integration test fails.
+                case "Include":
```
No need for the comment

```suggestion
                case "Include":
```

realvizu replied:
OK, removed the comment.

Rob-Hague commented on test/Renci.SshNet.Tests/Renci.SshNet.Tests.csproj:
```diff
@@ -3,6 +3,10 @@
   </PropertyGroup>
 
+  <PropertyGroup Condition=" '$(TargetFramework)' == 'net462' ">
+    <DefineConstants>$(DefineConstants);NET462_TARGET_FRAMEWORK</DefineConstants>
```
If we revert the FIPS test changes, then this is already unnecessary, but FYI the constant `NET462` is defined [by default](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives#conditional-compilation) so one can do `#if NET462` without any additional configuration

realvizu replied:
OK, removed it.

Rob-Hague reviewed (commented):
I don't think we should send channel EOF in `Write`. How about calling it only in `Dispose` and initialising `InputStream` lazily, something like:

```c#
var cmd = ...
using (cmd.InputStream) // This initialises InputStream
{
    myStream.CopyTo(cmd.InputStream);
} // This disposes InputStream and sends EOF

```

?

Commit:
Reverting changes made to unit tests unrelated to this PR.

Commit:
Moved ChannelInputStream's EOF sending from Write to Dispose. Replace SshCommand.InputStream with CreateInputStream to emphasise that a (disposable) resource is created here. EndExecute also closes the _inputStream to make sure that EOF is sent (in case the user forgot to dispose the input stream). Added more unit tests: sending the input one byte at a time, not disposing the input stream, calling CreateInputStream before BeginExecute or AfterEndExecute throws exception.

Commit:
Merge branch 'SshCommandInputStream' of https://github.com/realvizu/SSH.NET into SshCommandInputStream

realvizu commented:
> It looks good to me! But I need more time to test it!
> 
> Do you have a specific function in mind that you want to implement using this functionality? Did you see that it can be useful in [dotnet/Docker.DotNet#540](https://github.com/dotnet/Docker.DotNet/issues/540)

I had a very specific use case: I had to call some legacy system that is implemented as a set of SSH commands and one of the commands requires sending a file via input stream. These changes made it work.

realvizu commented:
Thanks a lot @WojciechNagorski and @Rob-Hague for the comments and suggestions!

- I reverted all the FIPS-related changes and the change made to SftpClientTest.Upload.cs.
- I moved the EOF sending to Dispose and added a unit test that validates that now multiple calls to Write works as expected.
- I've adapted @Rob-Hague 's suggestion about making InputStream lazily initialized with some modification. I wanted to make it more apparent that a resource is created here (that should be disposed after using), so I changed the InputStream from a property to a CreateInputStream method.
- Also, because the input stream uses the channel created in BeginExecute and closed in EndExecute, I added a check that throws invalid operation exception if CreateInputStream is called when there's no channel to use. Also added some unit tests for this.

Rob-Hague commented on src/Renci.SshNet/Common/ChannelInputStream.cs:
```diff
@@ -0,0 +1,220 @@
+    /// <summary>
+    /// ChannelInputStream is a one direction stream intended for channel data.
+    /// </summary>
+    public class ChannelInputStream : Stream
```
Since the constructor is internal, perhaps the type could just be internal unless there is a reason to make it public

realvizu replied:
Sure.

Rob-Hague commented on src/Renci.SshNet/Common/ChannelInputStream.cs:
```diff
@@ -0,0 +1,220 @@
+        /// </summary>
+        /// <param name="disposing"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
+        /// <remarks>
+        /// Disposing a <see cref="PipeStream"/> will interrupt blocking read and write operations.
```
Maybe remove this remark :-)

realvizu replied:
Sure.

Rob-Hague commented on src/Renci.SshNet/Common/ChannelInputStream.cs:
```diff
@@ -0,0 +1,220 @@
+                _isDisposed = true;
+
+                // Closing the InputStream requires sending EOF.
+                if (_totalPosition > 0 && _channel?.IsOpen == true)
```
This should guard on `disposing`

```suggestion
                if (disposing && _totalPosition > 0 && _channel?.IsOpen == true)
```

realvizu replied:
Right.

Rob-Hague commented on src/Renci.SshNet/SshCommand.cs:
```diff
@@ -64,6 +65,25 @@ public class SshCommand : IDisposable
+                throw new InvalidOperationException($"The input stream can be used only after calling BeginExecute and before calling EndExecute.");
+            }
+
+            _inputStream ??= new ChannelInputStream(_channel);
```
Perhaps it should throw an error if it's already not-null, instead of silently doing nothing?

realvizu replied:
Right.

Rob-Hague reviewed (approved):
Nice

Commit:
Fixing review comments.

realvizu commented:
Shall I do a squash+rebase onto develop before merging the PR?

Rob-Hague commented:
It will be merged with "Squash and merge" so it is not necessary

(it requires a CODEOWNERS approval before it can merge)

WojciechNagorski committed:
Merge branch 'develop' into SshCommandInputStream

WojciechNagorski committed:
Merge branch 'develop' into SshCommandInputStream

WojciechNagorski committed:
Merge branch 'develop' into SshCommandInputStream

WojciechNagorski commented:
There is a build problem:
> C:\projects\ssh-net\test\Renci.SshNet.IntegrationTests\SshClientTests.cs(37,33): error CS7036: There is no argument given that corresponds to the required parameter 'offset' of 'Stream.Write(byte[], int, int)' [C:\projects\ssh-net\test\Renci.SshNet.IntegrationTests\Renci.SshNet.IntegrationTests.csproj::TargetFramework=net48]

Rob-Hague commented on test/Renci.SshNet.IntegrationTests/SshClientTests.cs:
```diff
@@ -17,12 +17,99 @@ public SshClientTests()
+
+                using (var inputStream = command.CreateInputStream())
+                {
+                    inputStream.Write(inputByteArray);
```
```suggestion
                    inputStream.Write(inputByteArray, 0, inputByteArray.Length);
```

WojciechNagorski replied:
Thanks

Rob-Hague committed:
Fix build error after #1286

WojciechNagorski commented:
This issue has been fixed in the 2024.0.0 version.
